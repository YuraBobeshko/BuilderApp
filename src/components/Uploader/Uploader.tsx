import React from "react";
import JSZip from "jszip";
import { ITree, ITreeItem, ISetTree } from "../../types";

const Zip = new JSZip();

interface IUploader {
  setTree: ISetTree;
}

const Uploader = ({ setTree }: IUploader) => {
  return (
    <div>
      <input
        className="fileInput"
        type="file"
        onChange={(e) => onChange(e, setTree)}
      />
    </div>
  );
};

function onChange(e: React.ChangeEvent<HTMLInputElement>, setTree: ISetTree) {
  let reader = new FileReader();
  if (e.currentTarget.files == null) {
    throw new Error("Error finding e.currentTarget.files");
  }
  let file = e.currentTarget.files[0];

  reader.onloadend = async () => {
    const zip = await Zip.loadAsync(file);
    const listFolder = zip.files;
    const tree: ITree = [];

    zip.forEach((element, file) => {
      const path = listFolder[element].name.split("/").filter(Boolean);

      function createTree(path: string[], tree: ITree) {
        if (!path[0]) return tree;

        if (!tree.find((element: ITreeItem) => element.name === path[0])) {
          tree.push({
            name: path[0],
            children: [],
            type: file.dir
              ? "folder"
              : file.name.endsWith(".tsx")
              ? "component"
              : "file" || "select",
            text: file.async("string") || "",
          });
        }
        if (path.length)
          createTree(
            path.slice(1),
            tree?.find((element: ITreeItem) => element?.name === path[0])
              ?.children || []
          );
      }
      createTree(path, tree);
    });
    setTimeout(() => {
      setTree(tree);
    }, 500);
  };

  reader.readAsDataURL(file);
}

export default Uploader;
