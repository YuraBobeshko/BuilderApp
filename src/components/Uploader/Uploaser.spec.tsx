import 'jsdom-global/register';
import Uploader from "./Uploader";

// describe('Uploader component', () => {
// })
