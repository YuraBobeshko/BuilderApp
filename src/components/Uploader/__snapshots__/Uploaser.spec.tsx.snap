// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Uploader component should render Uploader 1`] = `
<div>
  <input
    className="fileInput"
    onChange={[Function]}
    type="file"
  />
</div>
`;
